src/torchjd/_autogram/_rev_gram_acc.py

```diff
@@ -0,0 +1,325 @@
+from collections import Counter, deque
+from collections.abc import Callable, Iterable, Sequence
+from typing import Any
+
+import torch
+from torch import Tensor, nn
+from torch.autograd.graph import GradientEdge, get_gradient_edge
+from torch.nn import Parameter
+from torchviz import make_dot
+
+from torchjd.aggregation import UPGrad
+from torchjd.aggregation._weighting_bases import PSDMatrix, Weighting
+
+# TODO: test with multi-input and / or multi-output modules
+# TODO: test with parameter re-use (intra-module, inter-module, and module reuse)
+# TODO: document the cases where it doesn't work: non-batched operations, free nn.Parameter used before other nn.Modules (should work but slowly).
+
+
+class GramianAccumulator:
+    def __init__(self):
+        self.gramian = None
+        self.jacobians = dict()
+        self.counter = Counter()
+
+    def track_parameters(self, tensors: Iterable[Tensor]) -> None:
+        self.counter.update(tensors)
+
+    def add_jacobian(self, tensor: Tensor, jacobian: Tensor) -> None:
+        if tensor in self.jacobians:
+            self.jacobians[tensor] += jacobian
+        else:
+            self.jacobians[tensor] = jacobian
+        self.counter.subtract([tensor])
+        if self.counter[tensor] == 0:
+            self._accumulate_jacobian(self.jacobians[tensor])
+            del self.counter[tensor]
+            del self.jacobians[tensor]
+
+    def _accumulate_jacobian(self, jacobian: Tensor) -> None:
+        if self.gramian is not None:
+            self.gramian.addmm_(jacobian, jacobian.T)
+        else:
+            self.gramian = torch.mm(jacobian, jacobian.T)
+
+
+def make_autogram_function(
+    module: nn.Module,
+    gramian_accumulator: GramianAccumulator,
+    args: Any,
+    outputs: list[Tensor],
+    named_params: dict[str, Tensor],
+) -> type[torch.autograd.Function]:
+
+    activated = True
+
+    class Function(torch.autograd.Function):
+        @staticmethod
+        def forward(xs: Any) -> None:
+            return xs
+
+        @staticmethod
+        def setup_context(*_):
+            pass
+
+        @staticmethod
+        def backward(ctx, *grad_outputs: Tensor):
+            nonlocal activated
+            if activated:
+                activated = False
+
+                def get_vjp(grad_outputs_j: tuple[Tensor, ...], *inputs_j) -> tuple[Tensor, ...]:
+                    inputs_j = [input_j.unsqueeze(0) for input_j in inputs_j]
+                    grad_outputs_j = [
+                        grad_output_j.unsqueeze(0) for grad_output_j in grad_outputs_j
+                    ]
+                    return _vjp_from_module(module, *inputs_j)(*grad_outputs_j)
+
+                jacobians = torch.vmap(get_vjp)(grad_outputs, *args)
+                assert len(jacobians) == 1
+                batch_size = outputs[0].shape[0]
+                for param_name, jacobian in jacobians[0].items():
```

Comment:
I think we could make `GramianAccumulator` take string as keys instead of tensors, this way we don't need `named_params` in this function.